Dockerfile

```diff
@@ -8,32 +8,32 @@ RUN set -ex; \
         python-dev \
         git
 
-RUN curl -fsSL -o dockerbins.tgz "https://download.docker.com/linux/static/stable/x86_64/docker-17.12.0-ce.tgz" && \
-    SHA256=692e1c72937f6214b1038def84463018d8e320c8eaf8530546c84c2f8f9c767d; \
-    echo "${SHA256}  dockerbins.tgz" | sha256sum -c - && \
-    tar xvf dockerbins.tgz docker/docker --strip-components 1 && \
-    mv docker /usr/local/bin/docker && \
-    chmod +x /usr/local/bin/docker && \
-    rm dockerbins.tgz
-
 # Python3 requires a valid locale
 RUN echo "en_US.UTF-8 UTF-8" > /etc/locale.gen && locale-gen
 ENV LANG en_US.UTF-8
 
 RUN useradd -d /home/user -m -s /bin/bash user
+RUN mkdir /code/ && chown -R user:user /code/
+RUN pip install tox==2.1.1
+
+# Install the docker cli, which is used for docker-compose exec by default
+ARG DOCKERBINS_VERSION=18.03.1
+ARG DOCKERBINS_SHA=0e245c42de8a21799ab11179a4fce43b494ce173a8a2d6567ea6825d6c5265aa
+RUN curl -fsSL -o dockerbins.tgz "https://download.docker.com/linux/static/stable/x86_64/docker-${DOCKERBINS_VERSION}-ce.tgz" && \
+    echo "${DOCKERBINS_SHA} dockerbins.tgz" | sha256sum -c - && \
+    tar xvf dockerbins.tgz docker/docker --strip-components 1 && \
+    mv docker /usr/local/bin/docker && \
+    chmod +x /usr/local/bin/docker && \
+    rm dockerbins.tgz
+
 WORKDIR /code/
 
-RUN pip install tox==2.1.1
+COPY --chown=user:user . /code/
```

Comment:
The point of doing individual ADDs like below is to avoid having to re-run the tox environment setup when making unrelated changes (like changes to test files). Ideally this is still something we'd want to keep.